### PR TITLE
JFW wip command multiplex (RFC, DNM!)

### PR DESCRIPTION
**Pull request by chardan**
** Request for Comments, please Do Not Merge! **

This branch introduces a mini-framework meant as a step toward unifying and standardizing the way that commands are multiplexed (I'd say dispatched, but that's taken) in Ceph (or at least mon). 

Because this is something that will be an investment to change after it begins to be used, I'd like to ask for comments at this point in the cycle. 

Some goals of this design include:

* compartmentalize commands (rather than having them be in one gigantic function)
* provide clearer semantics (explicit success/failure, reduce magical values, make the required
state explicit; reduce passing bare pointers, etc.)
* make adding new commands "easy"
  - centralize the parameters and naming conventions of variables
* increase visibility to make further improvements possible: 
  - see what parameters are or are not const
  - see the structure of commands in general
  - reduce overall complexity to see new patterns
* provide a generalized way to handle commands
  - core code and approach should be fairly flexible and reusable
* as command multiplexing is expected to be overshadowed by actual processing, I've proceeded under the assumption that some pointer indirection and lookup cost is acceptable

The basic idea:
There is a template or two (in common/command_multiplexer.hpp) that you can instantiate with whatever kind of context structure, etc. you want. Then, you bind commands into it. The normal behavior is for commands to be implemented via an interface, typically a user should just specify a dispatching tag (or tags) and implement the phase(s) they are interested in (eg. preprocess and prepare). 
The environment for commands is provided via a context structure that is automatically passed as a parameter to the command member functions. 

There are several additional things I'd like to draw attention to:
- In OSDMonitor, you can see a single function that handles the multiplexing.
- In MonmapMonitor, you can see what happens when all commands are migrated-- in particular, the context constructor is able to centralize all the details of initialization. The preprocess_command() and prepare_command() member functions become much simpler. (However, notice that in MonmapMonitor they are not currently dispatching into a single function because of a minor (and possibly undesirable) behavior difference, as noted in my code comment.)
- the context structures are meant to provide everything needed by any function, in no small part so we can see exactly what those things are. However, that does mean that there's overhead for constructing possibly unused things. The current assumption is that those are all sufficiently "cheap" as to have no significant effect on overall performance. We may need to look at other approaches, such as having data constructed inside functions that need them, however this may be more clear /after/ we migrate things.
- Another thought I had was to allow free functions, however providing a more traditional class hierarchy here allows us to help the user with a little more magic than we might get otherwise. 
- I realize some of the identifiers and namespaces are pretty verbose. I'm open to suggestions for new names, but I like to start with pretty explicit ideas.

Thank you very much for your time and constructive comments!

### CODE REVIEW

File: src/mon/Monitor.h [hunk 1]
```diff
@@ -855,7 +855,8 @@ class Monitor : public Dispatcher,
     }
   };
 
- private:
+ // JFW: we may want to come up with a solution based on friends:
```
Comment (chardan) [inline src/mon/Monitor.h]: Do folks want me to pursue this, or are we ok just making these types public?

File: src/mon/MonmapMonitor.h [hunk 2]
```diff
@@ -29,17 +32,74 @@
...
+ Monitor&                       mon;
+ MonMap*&                       monmap;         // aka "p", mon->monmap (sometimes mutated by side-effect)
+ MonMap&                        pending_map;
+ MMonCommand&                   mmon_command;   // aka "m"
+ MonSession*                    session;        // observing ptr
+ cmdmap_t                       cmdmap;
+ std::string                    prefix;
+ std::unique_ptr<Formatter>     fptr;           // JFW: why can't we have static Formatters..?
+ Formatter*                     formatter;      // aka f; observing ptr 
+ mutable std::stringstream      ss;
```
Comment (chardan) [inline src/mon/MonmapMonitor.h]: Part of the intent here is that we can /see/ what parameters are or are not const.

**Comment by chardan** [comment]
@jecluis Thanks for your valuable input! I've made most of the changes you suggested. I agree that documenting this is important, but I'd like to hear further comments and input on the overall design rather than documenting something that might change. If folks are feeling good about things, I'll be sure to do it before any merging. I'm happy to answer any questions in the meantime.

Another element that's going to take some thought is the access to the types in Monitor. Unfortunately, friendship can't be inherited in C++, so exposing those types another way is... going to be extremly gnarly and *probably* not worth it. Clearly the author of the original Monitor class wanted them private for a reason, but how open are we to exposing them? If there is a strong objection, there are ways to *kind of* get around this by using proxies in the base class (ceph::command), but this may be trouble. Thoughts?

**Comment by chardan** [comment]
@jecluis Ok... I've thought of another idea. It's actually even weirder, but it could work. Again, this is only arguably a good thing to try-- if there's not a good reason to keep those types private, I hope we can just change their access. However, that said, take a peek here (I'll have to tinker with it more tomorrow):
https://godbolt.org/g/93CLSJ

File: src/mon/MonmapMonitor.cc [hunk 3]
```diff
@@ -34,6 +33,96 @@ static ostream& _prefix(std::ostream *_dout, Monitor *mon) {
...
 }
+static ostream& _prefix(std::ostream *_dout, Monitor& mon) {
+  return _prefix(_dout, &mon);
+}
+
+#include <ostream>
+#include <sstream>
+#include <iostream>
+
+namespace ceph_cmds::monmapmon {
```
[the remark this answers is not in the document]
Comment (chardan) [reply src/mon/MonmapMonitor.cc]: I'll give this some more thought and change it.

**Comment by chardan** [comment]
@jecluis Ok-- let's change "probably" to "definitely" a bad idea. Not only does "a command is a Monitor" make no sense to me, it would also indirectly involve all the Paxos stuff and about a bazillion members, etc.. Unfortunately, that means that there's no probably no direct way around this. I'll keep thinking, but let's really consider if it's ok to expose the types instead.

**Comment by chardan** [comment]
@jecluis I've found /another/ way to do it, but it exposes /every/ private member to each command, including all the types. I doubt that's what we want, either...

**Comment by chardan** [comment]
...and, one more, which may actually be a reasonable answer. I'll try it tomorrow!
https://godbolt.org/g/S3M7z7

**Comment by chardan** [comment]
@jecluis Ok! I'm fairly happy with the Projection workaround. Basically, a proxy class that lets you filter in which private/protected members you want. It also would be possible to project those types into the command base class (by adding one more Level of Indirection (thanks, David Wheeler! ;-D), but I don't think that gains much and is a bit more complicated.

**Comment by chardan** [comment]
Jenkins retest this please.

**Comment by gregsfortytwo** [comment]
Can you write a bit comparing this to eg the src/mon/FSCommands.h system of handling things?

**Comment by chardan** [comment]
@gregsfortytwo Sure! This isn't all that different from the approach in FSCommands in most ways: if you squint a bit, they
share a lot of the same moving parts and functionality. That said, there are distinctions, and I'll try to highlight the
broad strokes.

The largest distinction is that this is mean to be less specific than FSCommands: rather than a handler class tied to a
particular module, this is a "micro-framework" meant to be customizable to pretty much anything we do in Mon that has the
prepare-preprocess pattern. (This is why there are two modules shown in the RFC: both to demonstrate the general nature and
to show a "complete" port and "incomplete" one.)

Similarly to FSCommands, users are meant to write function objects that are derived from a base class that generalizes commands
(an instantiation of class command<> in this framework, class FileSystemCommandHandler in FSCommands).

As part of this approach, the data and any member access from the owning class is decoupled as much as possible. This is captured
in a structure, which so far is called "state".[1] There's a lot to discuss about the implications of this choice-- which has both
advantages and disadvantages, but the key is that this type is decoupled from the rest of the framework.

The actual calling of functions and dispatching of a certain function based on a prefix trigger is moved into a centralized 
"command registry". Class MDSMonitor has "handlers", and in this proposal it would have "commands" (see class OSDMonitor). In
the former case, it's directly a list of pointers to function objects, in this framework it's an object that contains such a list
and also knows how to handle tags, etc.. This is handled by multiplex_command().

The function calls themselves have another layer around them (perhaps uncesssarily, but there are things I like about it, too), which
is that they return a distinct success or failure state along with a result code. Part of this is to step back from the idea of
setting flags with ranges having different meanings (I've seen a couple of places where I can't tell if a return code has been checked
correctly or not). This is surely worth further discussion, but for now that's the general point of it: to reduce the amount of customized
code needed for general command multiplexing.

Constructing the set of function objects is also pretty analagous. FileSystemCommandHandler::load() and register_comamnds() are
very simliar.

I've added further comments inside of command_multiplexer.h. Although there's a bit more complexity in the framework, and I'm a bit prone to
using verbose names especially when I'm writing something the first time around, from the perspective of someone wanting to add a new
function, I hope that it's straightforward:

struct osd_echo : public osdmonitor_command
{
 vector<string> tags() { return { "osd echo" }; }

 osdmonitor_command_result at_prepare(const osdmonitor_command_state& state, string_view cmd)
 {
    state.ss << cmd;
    return { osdmonitor_command_status::success, 0 };
 }
};

...is hopefully not too bad.

Hope that helps!

1. Currently all in a giant bucket. I don't see this as ultimately desirable, but for now it will at least let us decouple what
functions need from being able to access every single private member of complex classes like OSDMonitor. Besides allowing us to see
what data is actually used, another useful point is that this enables the state members to have consistent names by "force" rather than
"convention". Finally, it also makes it at least somewhat easier to see which parts of the state are actually mutated, and which can
be const. This comes at the cost of a bit of verbosity to the user (generally, I haven't found this to be that big of a deal) and some
nuisiance when porting existing code (usually not to bad, but occasionaly unnatural).

**Comment by liewegas** [comment]
This is slightly OT (and I haven't looked at this code yet :)), but one thing I think we should think about here are the usual behavior patterns we see in command implementations.  Read-only commands are simple; it's the updates that get complicated. Roughly

1. No-op.  All commands are idempotent, so if the current state matches what the command wants then return immediately.
2. Update.  Prepare a change, reply after it commits to paxos.
3. Wait.  The current (committed) state doesn't match what we want, but there is some already-pending state change that may conflict with what we want (e.g., because the proposed change is already pending, or the value we want to set is already being changed to something else first, etc).  In this case, we block until the pending proposal finishes, then restart the command (starting with preprocess).
4. Error.  We can't do what the command wants, return an error.

It's the (3) case that we often ignore or implement in a sloppy fashion.  Maybe it's trivial to make the osdmonitor_command_result be a 3-way enum and that's all that's needed?  Usually the condition checks for 3 are tedious and semi-redundant (check pending_inc.foo in addition to foo). Probably no way around that tho?

**Comment by chardan** [comment]
@liewegas This is the /perfect/ time to consider those things! It's easy enough to make the enum ternary (now that there is an enum). As for handling those cases... maybe once things are split apart a bit more it will be easier to see the right pattern. Please note that I may not have gotten it right in the handling I've written here (if not, hopefully straightforward to change).

I'll try to get this passing checks again soon, likely next week. (I seem to missing some work, probably I messed up a merge.) In any case, though, so it's still important to get input.

**Comment by chardan** [comment]
I think it will be easier to consider the basic moving parts in a context where there's not as much code ported, so I'll close this and reference the work and discussions that are already here.

Please refer-to the now opened:
https://github.com/ceph/ceph/pull/24093

**Pull request closed without merging**